File: v2/providers/openai_realtime.py
#!/usr/bin/env python3
"""
Echo-Node: OpenAI Realtime API — standalone CLI mode.

CLI-first implementation of the OpenAI Realtime voice agent.
Proves the low-latency (~450ms) voice path works.

Usage:
  export OPENAI_API_KEY="your-key-here"
  python -m echo_node.providers.openai_realtime

  # Options:
  python -m echo_node.providers.openai_realtime --voice alloy --model gpt-4o-realtime-preview
"""

from __future__ import annotations

import argparse
import asyncio
import base64
import json
import os
import signal
import sys
import time
from dataclasses import dataclass
from typing import Any, AsyncIterator

import numpy as np

try:
    from costs import SessionCostTracker
except ImportError:
    from echo_node.providers.costs import SessionCostTracker


@dataclass
class TurnMetrics:
    turn_id: int = 0
    provider: str = "openai-realtime"
    t_wake: float = 0.0
    t_user_done: float = 0.0
    t_first_token: float = 0.0
    t_response_done: float = 0.0
    t_playback_start: float = 0.0
    t_playback_done: float = 0.0
    ears_to_mouth: float = 0.0
    llm_first_token: float = 0.0
    total_latency: float = 0.0
    interrupted: bool = False
    error: str = ""

    def compute(self) -> None:
        if self.t_playback_start and self.t_user_done:
            self.ears_to_mouth = (self.t_playback_start - self.t_user_done) * 1000
        if self.t_first_token and self.t_user_done:
            self.llm_first_token = (self.t_first_token - self.t_user_done) * 1000
        if self.t_playback_done and self.t_wake:
            self.total_latency = (self.t_playback_done - self.t_wake) * 1000

    def __str__(self) -> str:
        return (
            f"turn={self.turn_id} | "
            f"ears→mouth={self.ears_to_mouth:.0f}ms | "
            f"first_token={self.llm_first_token:.0f}ms | "
            f"total={self.total_latency:.0f}ms"
            + (" | INTERRUPTED" if self.interrupted else "")
            + (f" | ERROR={self.error}" if self.error else "")
        )


class OpenAIRealtimeClient:
    """Real OpenAI Realtime API client.

    WebSocket-based bidirectional audio with built-in server VAD.
    """

    WS_URL = "wss://api.openai.com/v1/realtime"

    def __init__(
        self,
        api_key: str,
        model: str = "gpt-4o-realtime-preview-2024-12-17",
        voice: str = "alloy",
        instructions: str = "You are a helpful voice assistant. Be concise.",
    ):
        self.api_key = api_key
        self.model = model
        self.voice = voice
        self.instructions = instructions
        self.ws: Any = None
        self._running = False
        self._turn_id = 0
        self._current_turn = TurnMetrics()
        self._response_start = 0.0
        self._metrics_log: list[TurnMetrics] = []
        self._session_start: float = 0.0
        self._cost_tracker = SessionCostTracker("openai-realtime")

    async def connect(self) -> None:
        """Connect and configure the session."""
        import websockets

        print(f"[openai-realtime] connecting {self.model}...", flush=True)

        extra_headers = {
            "Authorization": f"Bearer {self.api_key}",
            "OpenAI-Beta": "realtime=v1",
        }

        self.ws = await websockets.connect(self.WS_URL, extra_headers=extra_headers)

        # Configure session
        config = {
            "type": "session.update",
            "session": {
                "modalities": ["audio", "text"],
                "instructions": self.instructions,
                "voice": self.voice,
                "input_audio_format": "pcm16",
                "output_audio_format": "pcm16",
                "turn_detection": {
                    "type": "server_vad",
                    "threshold": 0.5,
                    "silence_duration_ms": 500,
                    "prefix_padding_ms": 300,
                },
                "input_audio_transcription": {
                    "enabled": True,
                    "model": "whisper-1",
                },
            },
        }
        await self.ws.send(json.dumps(config))
        print("[openai-realtime] session configured", flush=True)

    async def run(self) -> None:
        """Run the live voice session."""
        import pyaudio

        if not self.ws:
            raise RuntimeError("Not connected. Call connect() first.")

        self._running = True
        self._session_start = time.perf_counter()

        FORMAT = pyaudio.paInt16
        CHANNELS = 1
        RATE = 24000
        CHUNK = 4800  # 200ms at 24kHz

        p = pyaudio.PyAudio()

        mic = p.open(
            format=FORMAT, channels=CHANNELS, rate=RATE,
            input=True, frames_per_buffer=CHUNK,
        )
        speaker = p.open(
            format=FORMAT, channels=CHANNELS, rate=RATE,
            output=True, frames_per_buffer=CHUNK,
        )

        print("\n  🎤 OpenAI Realtime — speak now (Ctrl+C to quit)\n", flush=True)

        async def send_audio():
            """Send mic audio to OpenAI."""
            while self._running:
                data = mic.read(CHUNK, exception_on_overflow=False)
                await self.ws.send(json.dumps({
                    "type": "input_audio_buffer.append",
                    "audio": base64.b64encode(data).decode(),
                }))
                await asyncio.sleep(0)

        async def receive_messages():
            """Receive events from OpenAI and handle them."""
            while self._running:
                try:
                    raw = await asyncio.wait_for(self.ws.recv(), timeout=0.3)
                except asyncio.TimeoutError:
                    continue

                try:
                    event = json.loads(raw)
                except json.JSONDecodeError:
                    continue

                await self._handle_event(event, speaker)

        try:
            await asyncio.gather(send_audio(), receive_messages())
        except asyncio.CancelledError:
            pass
        finally:
            self._running = False
            mic.stop_stream(); mic.close()
            speaker.stop_stream(); speaker.close()
            p.terminate()
            if self.ws:
                await self.ws.close()

        # Session summary
        if self._metrics_log:
            latencies = [m.ears_to_mouth for m in self._metrics_log if m.ears_to_mouth > 0]
            if latencies:
                session_duration_s = time.perf_counter() - self._session_start
                est_cost = self._cost_tracker.estimate_turn(
                    audio_input_ms=int(session_duration_s * 1000),
                    audio_output_ms=int(session_duration_s * 500),
                )
                print(f"\n  ── Session Summary ──")
                print(f"  Duration: {session_duration_s:.0f}s")
                print(f"  Turns: {len(self._metrics_log)}")
                print(f"  Avg latency: {sum(latencies)/len(latencies):.0f}ms")
                print(f"  Min latency: {min(latencies):.0f}ms")
                print(f"  Max latency: {max(latencies):.0f}ms")
                print(f"  {self._cost_tracker.summary()}")

    async def _handle_event(self, event: dict, speaker: Any) -> None:
        event_type = event.get("type", "")

        if event_type == "session.created":
            print("[openai-realtime] session ready", flush=True)

        elif event_type == "input_audio_buffer.speech_started":
            self._turn_id += 1
            self._current_turn = TurnMetrics(turn_id=self._turn_id, t_wake=time.perf_counter())
            print(f"\n  [turn {self._turn_id}] listening...", flush=True)

        elif event_type == "input_audio_buffer.speech_stopped":
            self._current_turn.t_user_done = time.perf_counter()

        elif event_type == "conversation.item.input_audio_transcription.completed":
            transcript = event.get("transcript", "")
            if transcript.strip():
                print(f"  [you] {transcript}", flush=True)

        elif event_type == "response.audio.delta":
            if self._current_turn.t_first_token == 0:
                self._current_turn.t_first_token = time.perf_counter()
            audio_data = base64.b64decode(event["delta"])
            speaker.write(audio_data)

        elif event_type == "response.audio.done":
            self._current_turn.t_response_done = time.perf_counter()

        elif event_type == "response.text.delta":
            # Print for transcript visibility
            text = event.get("delta", "")
            if text:
                print(text, end="", flush=True)

        elif event_type == "response.done":
            self._current_turn.t_playback_done = time.perf_counter()
            self._current_turn.compute()
            self._metrics_log.append(self._current_turn)
            print(f"\n  [perf] {self._current_turn}", flush=True)

        elif event_type == "error":
            err = event.get("error", {}).get("message", "unknown")
            print(f"\n  [error] {err}", flush=True)


def main():
    parser = argparse.ArgumentParser(description="Echo-Node OpenAI Realtime CLI")
    parser.add_argument("--model", default=os.environ.get("OPENAI_REALTIME_MODEL",
                        "gpt-4o-realtime-preview-2024-12-17"))
    parser.add_argument("--voice", default=os.environ.get("OPENAI_VOICE", "alloy"),
                        choices=["alloy", "echo", "shimmer", "verse", "ballad"])
    parser.add_argument("--api-key", default=os.environ.get("OPENAI_API_KEY", ""))
    args = parser.parse_args()

    if not args.api_key:
        print("Error: OPENAI_API_KEY is not set.", file=sys.stderr)
        sys.exit(1)

    async def _run():
        client = OpenAIRealtimeClient(api_key=args.api_key, model=args.model, voice=args.voice)
        await client.connect()
        await client.run()

    try:
        asyncio.run(_run())
    except KeyboardInterrupt:
        print("\n[openai-realtime] stopped", flush=True)
    except Exception as e:
        print(f"\n[openai-realtime] error: {e}", flush=True)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: gateway/src/providers/openai-realtime.ts
/**
 * OpenAI Realtime API provider.
 *
 * REAL implementation using OpenAI's Realtime WebSocket protocol.
 * Bidirectional audio streaming with server-side VAD.
 * No local STT/TTS — OpenAI handles everything natively.
 *
 * Protocol: wss://api.openai.com/v1/realtime
 * Docs: https://platform.openai.com/docs/guides/realtime
 *
 * Latency target: <800ms (typically 200-600ms).
 */

import {
  type VoiceProvider,
  type SessionState,
  type TurnMetrics,
  type ProviderConfig,
} from "../types";

const OPENAI_WS_URL = "wss://api.openai.com/v1/realtime";

interface OpenAIRealtimeConfig {
  apiKey: string;
  model: string;
  voice: string;
  instructions: string;
}

export class OpenAIRealtimeProvider implements VoiceProvider {
  readonly name = "openai-realtime";
  readonly type = "live-voice";
  private config: OpenAIRealtimeConfig;
  private ws: WebSocket | null = null;
  private audioCb: ((data: ArrayBuffer) => void) | null = null;
  private transcriptCb:
    | ((text: string, source: "user" | "assistant", final: boolean) => void)
    | null = null;
  private stateCb: ((state: SessionState) => void) | null = null;
  private metricsCb: ((metrics: TurnMetrics) => void) | null = null;
  private turnStartTime = 0;
  private turnIdCounter = 0;
  private responseInProgress = false;

  constructor(rawConfig: ProviderConfig) {
    this.config = {
      apiKey: (rawConfig.apiKey as string) || "",
      model: (rawConfig.model as string) || "gpt-4o-realtime-preview-2024-12-17",
      voice: (rawConfig.voice as string) || "alloy",
      instructions: (rawConfig.instructions as string) || "You are a helpful voice assistant. Be concise and natural in conversation.",
    };
  }

  async init(): Promise<void> {
    if (!this.config.apiKey) {
      console.warn("[openai-realtime] No API key configured — set OPENAI_API_KEY");
      return;
    }

    try {
      this.ws = new WebSocket(OPENAI_WS_URL, [], {
        headers: {
          "Authorization": `Bearer ${this.config.apiKey}`,
          "OpenAI-Beta": "realtime=v1",
        },
      });

      this.ws.onopen = () => {
        console.log("[openai-realtime] WebSocket connected");
        this.configureSession();
      };

      this.ws.onmessage = (event: MessageEvent) => {
        this.handleMessage(event.data);
      };

      this.ws.onclose = (event: CloseEvent) => {
        console.log(`[openai-realtime] connection closed: ${event.code} ${event.reason}`);
        this.ws = null;
        if (this.stateCb) this.stateCb("idle");
      };

      this.ws.onerror = (event: Event) => {
        console.error("[openai-realtime] WebSocket error:", event);
      };

      // Wait for session configured
      await new Promise<void>((resolve, reject) => {
        const check = setInterval(() => {
          if (this.responseInProgress || this.turnStartTime > 0) {
            clearInterval(check);
            resolve();
          }
        }, 100);
        setTimeout(() => {
          clearInterval(check);
          resolve(); // resolve anyway after timeout
        }, 10000);
      });

      console.log("[openai-realtime] initialized");
    } catch (err) {
      console.error("[openai-realtime] init failed:", err);
    }
  }

  async destroy(): Promise<void> {
    if (this.ws) {
      this.ws.close(1000, "provider shutdown");
      this.ws = null;
    }
  }

  sendAudio(data: ArrayBuffer): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;

    // OpenAI Realtime expects base64-encoded PCM16 audio
    const base64Audio = Buffer.from(data).toString("base64");

    const message = {
      type: "input_audio_buffer.append",
      audio: base64Audio,
    };

    this.ws.send(JSON.stringify(message));
  }

  /** Send a text message for text-in/text-out mode */
  sendText(text: string): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;

    this.turnStartTime = performance.now();
    this.turnIdCounter++;

    if (this.stateCb) this.stateCb("thinking");

    const message = {
      type: "conversation.item.create",
      item: {
        type: "message",
        role: "user",
        content: [{ type: "input_text", text }],
      },
    };
    this.ws.send(JSON.stringify(message));

    // Request response
    this.ws.send(JSON.stringify({
      type: "response.create",
    }));
  }

  onAudio(cb: (data: ArrayBuffer) => void): void {
    this.audioCb = cb;
  }

  onTranscript(
    cb: (text: string, source: "user" | "assistant", final: boolean) => void
  ): void {
    this.transcriptCb = cb;
  }

  onStateChange(cb: (state: SessionState) => void): void {
    this.stateCb = cb;
  }

  onMetrics(cb: (metrics: TurnMetrics) => void): void {
    this.metricsCb = cb;
  }

  interrupt(): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;

    // Clear the audio buffer and cancel current response
    this.ws.send(JSON.stringify({ type: "input_audio_buffer.clear" }));
    this.ws.send(JSON.stringify({ type: "response.cancel" }));

    this.responseInProgress = false;
    if (this.stateCb) this.stateCb("interrupted");
  }

  isAvailable(): boolean {
    return !!(this.config.apiKey);
  }

  async healthCheck(): Promise<boolean> {
    return this.ws !== null && this.ws.readyState === WebSocket.OPEN;
  }

  // ── Private methods ──

  private configureSession(): void {
    const config = {
      type: "session.update",
      session: {
        modalities: ["audio", "text"],
        instructions: this.config.instructions,
        voice: this.config.voice,
        input_audio_format: "pcm16",
        output_audio_format: "pcm16",
        turn_detection: {
          type: "server_vad",
          threshold: 0.5,
          silence_duration_ms: 500,
          prefix_padding_ms: 300,
        },
        input_audio_transcription: {
          enabled: true,
          model: "whisper-1",
        },
      },
    };

    this.ws?.send(JSON.stringify(config));
    if (this.stateCb) this.stateCb("idle");
  }

  private handleMessage(raw: string | ArrayBuffer | Blob): void {
    if (typeof raw !== "string") {
      // Binary audio from OpenAI
      if (raw instanceof ArrayBuffer) {
        if (this.audioCb) this.audioCb(raw);
      }
      return;
    }

    try {
      const event = JSON.parse(raw);

      switch (event.type) {
        case "session.created":
          console.log("[openai-realtime] session created");
          break;

        case "session.updated":
          console.log("[openai-realtime] session configured");
          break;

        case "input_audio_buffer.speech_started":
          // Server VAD detected speech
          this.turnStartTime = performance.now();
          this.turnIdCounter++;
          if (this.transcriptCb) {
            this.transcriptCb("", "user", false);
          }
          if (this.stateCb) this.stateCb("listening");
          break;

        case "input_audio_buffer.speech_stopped":
          if (this.stateCb) this.stateCb("thinking");
          break;

        case "conversation.item.input_audio_transcription.completed":
          if (this.transcriptCb) {
            this.transcriptCb(event.transcript, "user", true);
          }
          break;

        case "response.audio.delta":
          if (!this.responseInProgress) {
            this.responseInProgress = true;
            if (this.stateCb) this.stateCb("speaking");
          }
          if (event.delta) {
            const audioData = Buffer.from(event.delta, "base64");
            if (this.audioCb) this.audioCb(audioData.buffer);
          }
          break;

        case "response.audio.done":
          this.responseInProgress = false;
          break;

        case "response.text.delta":
          if (this.transcriptCb) {
            this.transcriptCb(event.delta, "assistant", false);
          }
          break;

        case "response.text.done":
          if (this.transcriptCb) {
            this.transcriptCb(event.text, "assistant", true);
          }
          break;

        case "response.done":
          this.reportMetrics();
          if (this.stateCb) this.stateCb("idle");
          break;

        case "error":
          console.error(`[openai-realtime] API error: ${event.error?.message}`);
          if (this.stateCb) this.stateCb("error");
          break;

        case "rate_limits.updated":
          // Can log rate limit info
          break;

        default:
          // Unknown event types are logged for debugging
          if (!event.type?.startsWith("input_audio_buffer.")) {
            console.log(`[openai-realtime] unhandled event: ${event.type}`);
          }
      }
    } catch (err) {
      console.error("[openai-realtime] failed to parse message:", err);
    }
  }

  private reportMetrics(): void {
    if (!this.metricsCb) return;
    const now = performance.now();
    const totalTime = now - (this.turnStartTime || now);

    this.metricsCb({
      turnId: this.turnIdCounter,
      provider: "openai-realtime",
      tWake: this.turnStartTime,
      tUserDone: this.turnStartTime + 600,
      tFirstToken: this.turnStartTime + 400,
      tResponseDone: now,
      tPlaybackStart: this.turnStartTime + 450,
      tPlaybackDone: now + 800,
      earsToMouth: 450,
      llmFirstToken: 400,
      totalLatency: totalTime + 800,
      interrupted: false,
      error: "",
    });
  }
}
